libbpf-rs/tests/bin/src/cgroup-fw.bpf.c

```diff
@@ -0,0 +1,153 @@
+struct {
+    __uint(type, BPF_MAP_TYPE_HASH);
+    __uint(max_entries, 1000);
+    __type(key, __u32);
+    __type(value, _Bool);
+} ingress_blacklist SEC(".maps");
+
+/* store L3 blacklist rules for egress */
+struct {
+    __uint(type, BPF_MAP_TYPE_HASH);
+    __uint(max_entries, 1000);
+    __type(key, __u32);
+    __type(value, _Bool);
+} egress_blacklist SEC(".maps");
+
+/* store L4 blacklist rules for ingress */
+struct {
+    __uint(type, BPF_MAP_TYPE_HASH);
+    __uint(max_entries, 1000);
+    __type(key, skt);
+    __type(value, _Bool);
+} ingress_l4_blacklist SEC(".maps");
+
+/* store L4 blacklist rules for egress */
+struct {
+    __uint(type, BPF_MAP_TYPE_HASH);
+    __uint(max_entries, 1000);
+    __type(key, skt);
+    __type(value, _Bool);
+} egress_l4_blacklist SEC(".maps");
+
+/* store network flow logs for container */
+struct {
+    __uint(type, BPF_MAP_TYPE_QUEUE);
+    __uint(max_entries, 1000);
+    __uint(key, 0);
+    __type(value, pkt);
+} data_flow SEC(".maps");
+
+/* apply saved rules to ingress/egress packets, drop the packet if match */
+static inline int filter_packet(struct __sk_buff *skb, bool isIngress) {
+    void *data = (void*)(long)skb->data;
+    void *data_end = (void*)(long)skb->data_end;
+
+    struct iphdr *iphd = data;
+    __u32 iphdr_len = sizeof(struct iphdr);
+    // avoid verifier's complain
+    if (data + iphdr_len > data_end)
+        return 1;
+
+    skt s;
+    // initialize padding to avoid verifier's complain
+    __builtin_memset(&s, 0, sizeof(skt));
+
+    pkt p;
+    // initialize padding to avoid verifier's complain
+    __builtin_memset(&p, 0, sizeof(pkt));
+
+    p.saddr = iphd->saddr;
+    p.daddr = iphd->daddr;
+    p.proto = iphd->protocol;
+    s.proto = iphd->protocol;
+
+    if (iphd->protocol == IPPROTO_TCP) {
+        struct tcphdr *tcphd = data + iphdr_len;
+        __u32 tcphdr_len = sizeof(struct tcphdr);
+        // avoid verifier's complain
+        if ((void *)tcphd + tcphdr_len > data_end)
+            return 1;
+
+        p.sport = tcphd->source;
+        p.dport = tcphd->dest;
+    }
+
+    if (iphd->protocol == IPPROTO_UDP) {
+        struct udphdr *udphd = data + iphdr_len;
+        __u32 udphdr_len = sizeof(struct udphdr);
+        // avoid verifier's complain
+        if ((void *)udphd + udphdr_len > data_end)
+            return 1;
+
+        p.sport = udphd->source;
+        p.dport = udphd->dest;
+    }
+
+    bool isBannedL3;
+    bool isBannedL4;
+    if (isIngress) {
+        bpf_printk("Ingress from %lu",iphd->saddr);
+	s.addr = p.saddr;
+	// we don't care source port
+	s.port = p.dport;
+	isBannedL3 = bpf_map_lookup_elem(&ingress_blacklist, &iphd->saddr);
+	isBannedL4 = bpf_map_lookup_elem(&ingress_l4_blacklist, &s);
+    }else{
+        bpf_printk("Egress to %lu",iphd->daddr);
+	s.addr = p.daddr;
+	s.port = p.dport;
+	isBannedL3 = bpf_map_lookup_elem(&egress_blacklist, &iphd->daddr);
+	isBannedL4 = bpf_map_lookup_elem(&egress_l4_blacklist, &s);
+    }
```

Comment:
Please don't mix tabs and spaces. Also, please fix formatting around `else`.

Reply:
tks to point out ! I removed this part as you advised below, but I'll apply a formatter in another repo where this file  comes from.